Merge changes from dd-trace-java 0.51.0

trask opened the pull request:
https://github.com/DataDog/dd-trace-java/releases/tag/v0.51.0

I went a different route on this merge and cherry-picked the relevant PRs, dealing with merge conflicts along the way.

iNikem commented on instrumentation/cassandra/cassandra-3.0/src/main/java/io/opentelemetry/auto/instrumentation/cassandra/v3_0/TracingSession.java:
```diff
@@ -45,7 +46,9 @@
 
-  private final ExecutorService executorService = Executors.newCachedThreadPool();
+  private static final ExecutorService EXECUTOR_SERVICE =
+      Executors.newCachedThreadPool(new DaemonThreadFactory("dd-cassandra-session-executor"));
```
Remove `dd-` prefix?

iNikem commented on instrumentation/lettuce/lettuce-5.0/lettuce-5.0.gradle:
```diff
@@ -28,7 +28,7 @@ dependencies {
 
   testCompile group: 'com.github.kstyrc', name: 'embedded-redis', version: '0.6'
   testCompile group: 'io.lettuce', name: 'lettuce-core', version: '5.0.0.RELEASE'
+  testCompile project(':instrumentation:reactor-3.1')
```
Is this needed for all tests or for `latestDepTest` only?

devinsba replied:
All, the earliest versions of Lettuce 5 use project reactor

iNikem commented on instrumentation/okhttp/okhttp-2.2/okhttp-2.2.gradle:
```diff
@@ -0,0 +1,41 @@
+  compileOnly(group: 'com.squareup.okhttp', name: 'okhttp', version: '2.2.0')
+
+  compile(project(':auto-tooling')) {
+    exclude module: 'okhttp'
```
I don't see that `auto-tooling` depends on `okhttp`

iNikem commented on instrumentation/okhttp/okhttp-2.2/okhttp-2.2.gradle:
```diff
@@ -0,0 +1,41 @@
+    exclude module: 'okhttp'
+  }
+  testCompile(project(':instrumentation:java-concurrent')) {
+    exclude module: 'okhttp'
```
Same here?

iNikem commented on instrumentation/okhttp/okhttp-2.2/src/test/groovy/OkHttp2AsyncTest.groovy:
```diff
@@ -0,0 +1,61 @@
+      .headers(Headers.of(HeadersUtil.headersToArray(headers)))
+      .build()
+
+    AtomicReference<Response> responseRef = new AtomicReference()
```
Maybe easier to use one `CompletableFuture` instead of two `AtomicReference` and a `CountDownLatch`? Callback will call either `complete` or `completeExceptionally` and test can end in just one line `return future.get(10, SECONDS)`.

trask replied:
Ah, this didn't end up working out because the Java 7 tests failed, oh well, it was a nice idea 👍

devinsba commented on instrumentation/lettuce/lettuce-5.0/src/test/groovy/LettuceReactiveClientTest.groovy:
```diff
@@ -287,4 +290,121 @@ class LettuceReactiveClientTest extends AgentTestRunner {
+    when:
+    runUnderTrace("test-parent") {
+      reactiveCommands.set("a", "1")
+        .then(reactiveCommands.get("a")) // The get here is ending up in another trace
```
I should have removed these comments. They no longer apply now that the tests pass

trask commented:
Thanks @devinsba @iNikem! I addressed your comments. I will rebase the "fixup" commits and force push in a bit to clean up history since we are not squashing these "merge from dd-trace-java" PRs.

iNikem commented on instrumentation/okhttp/okhttp-2.2/src/test/groovy/OkHttp2AsyncTest.groovy:
```diff
@@ -0,0 +1,46 @@
+        future.completeExceptionally(e)
+      }
+    })
+    return future.get().code()
```
timeout on `get`?
